src/utilities/Dealer.cpp

```diff
@@ -19,93 +23,126 @@ Dealer::DealerFlag::DealerFlag(DealerFlagTitle title, DealerFlagPriority priorit
+                        output.insert({roleNames[current.originalRolesIndex[j]], allRobots[current.originalIDsIndex.back()]});
+                    }
+                }
+                if (output.size() == allRobots.size()) return output;               // case if there are less then 11 bots to distribute
+                distribute_remove(current,indexRoles,indexID,scores);
+            }
+        }
+    }
+    return output;
 }
 
-/* assignment now has the robot index in allRobots (not id) at the role index, and is ordered according to the roleNames
- * for example: assignment[0] = 2 // index
- * and roleNames[0] = role_1
- * robot_id = allRobots[index]
- * --> we can therefore make a map of <rolename, robot_id>
- */
-std::unordered_map<std::string, v::RobotView> Dealer::mapFromAssignments(const std::vector<v::RobotView> &allRobots, const Dealer::FlagMap &flagMap,
-                                                                         const std::vector<int> &assignment) const {
-    vector<string> orderedRoleNames;
+void Dealer::distribute_init(std::vector<int>& indexRoles, std::vector<int>& indexID, std::vector<std::string>& roleNames, const Dealer::FlagMap &flagMap) {
+    indexRoles.reserve(flagMap.size());
+    indexID.reserve(flagMap.size());
+    roleNames.reserve(flagMap.size());
+    for (int i = 0; i < flagMap.size(); i++) {
+        indexRoles.push_back(i);
+        indexID.push_back(i);
+    }
     for (auto const &[roleName, dealerFlags] : flagMap) {
-        orderedRoleNames.push_back(roleName);
+        roleNames.push_back(roleName);
     }
-    unordered_map<string, v::RobotView> result;
-    for (int i = 0; i < orderedRoleNames.size(); i++) {
-        if (assignment[i] != -1) {
-            result.insert({orderedRoleNames[i], allRobots[assignment[i]]});
+}
+
+void Dealer::distribute_remove(DealerDistribute& current, std::vector<int>& indexRoles, std::vector<int>& indexID, std::vector<RoleScores>& scores){
+    std::sort(current.currentIDs.begin(), current.currentIDs.end());                    // Sort to delete from back to front
+    // Delete assigned roles and robots from score
+    for (int j = current.currentRoles.size() - 1; j >= 0; j--) {
+        scores.erase(scores.begin() + current.currentRoles[j]);                 // remove role from score (col)
+        indexRoles.erase(indexRoles.begin() + current.currentRoles[j]);         // remove from index list
+    }
+    for (int col = 0; col < scores.back().robotScores.size(); col++) {        // go through each score role (row)
+        for (int j = current.currentIDs.size() - 1; j >= 0; j--) {
+            scores[col].robotScores.erase(scores[col].robotScores.begin() + current.currentIDs[j]);     // remove the robot
         }
     }
-    return result;
+    for (int j = current.currentIDs.size() - 1; j >= 0; j--) {
+        indexID.erase(indexID.begin() + current.currentIDs[j]);                 // remove from index list
+    }
```

Comment:
I dislike this function, you should probably refactor it a bit.

I don't understand what you are doing line 81-83. It seems you're assuming these vectors have the same length but it's not super clear to me. Similarly, you're doing weird things in other places.

Secondly, by calling erase manually every time,you are performing a lot resizes (even though you do it back to front). You can just delete the entire range of elements.

Thirdly, use iterators! This makes it much more clear what you're iterating over. I suggest you look into .rend() and .rbegin() of std::vector<>. If you don't know what an iterator is, just let me know and I'll explain you.

Reply:
In 81-83 i am removing the roles that have been given away already. I.e.:
Start: Currently:
scores (A represent the index of the role, a1 represent score for robot1 for role A, this is scores[A] = a1,a2,a3,a4)
[0] A -> a1 a2 a3 a4 | 2
[1] B -> b1 b2 b3 b4 | 0
[2] C -> c1 c2 c3 c4  | 2
[3] D -> d1 d2 d3 d4 | 1
Now you want to give away roles A and C as they have the highest priority, currentScores are:
[0] A -> a1 a2 a3 a4
[1] C -> c1 c2 c3 c4
C has [2] in the real matrix and therefore this needs to be remembered. When the roles are given away (to robot 2 and 4 in this example) you get, assignments:
[0] 2
[1] 4
linking A to 2 and C to 4.
Finally you want to remove given away roles and robots so you want:
[0] B -> b1 b3 | 0
[1] D -> d1 d3 | 1
That's why there are so many vectors to keep in mind the index of the original and current role and id...
I agree, i did like to refactor this aswell, however i am not sure how to effectivily do that with 2D vectors. And not sure if the result will be more efficient (more indexes to go over with i.e. NaN elements for the removed roles and Id's).

Reply:
Is remove better than erase? 
https://stackoverflow.com/questions/4115279/most-efficient-way-of-erasing-deleting-multiple-stdvector-elements-while-retai